agent/command/s3_evergreen_credentials.go

```diff
@@ -27,16 +30,21 @@ type evergreenCredentialProvider struct {
 // createEvergreenCredentials creates a new evergreenCredentialProvider. It supports
 // long operations or operations that might need to request new credentials during
 // the operation (e.g. multipart bucket uploads).
-func createEvergreenCredentials(comm client.Communicator, taskData client.TaskData, roleARN string, updateExternalID func(string)) *evergreenCredentialProvider {
+func createEvergreenCredentials(comm client.Communicator, taskData client.TaskData, existingCredentials *aws.Credentials, roleARN string, updateExternalID func(string)) *evergreenCredentialProvider {
 	return &evergreenCredentialProvider{
-		comm:             comm,
-		taskData:         taskData,
-		roleARN:          roleARN,
-		updateExternalID: updateExternalID,
+		comm:                comm,
+		taskData:            taskData,
+		existingCredentials: existingCredentials,
+		roleARN:             roleARN,
+		updateExternalID:    updateExternalID,
 	}
 }
 
 func (p *evergreenCredentialProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	if p.existingCredentials != nil && p.existingCredentials.AccessKeyID != "" && !p.existingCredentials.Expired() {
```

Comment:
Is there any chance that the existing credentials could expire while the request is going on (like if the credentials are valid only for the next 1 second)? Or does the AWS SDK transparently call `Retrieve` again to refresh the credentials once they've expired?

Reply:
I've looked through their source code and it's hard to say. Every credential provider gets wrapped in their cache, and then passed through 5 different layers of whatever.

Looking more, it looks like their ExpiryWindow logic is meant to handle this- which in this case won't work since the existingCredentials might be given to the cache (which is where the ExpiryWindow logic lies) is already expired.

Reply:
I've added a manual expiry window to account for this